### ISSUE TITLE
New parser: URL

### ISSUE THREAD

**Issue by archiif** [kellyjonbrazil/jc#263]
A parser for URL strings would be nice.
Some possible examples:
```
$ echo "http://example.com/test/path?q1=foo&q2=bar#frag" | jc -p --url
{
  "scheme": "http",
  "host": "example.com",
  "path": ["test", "path"],
  "query": { "q1": "foo", "q2": "bar" },
  "fragment": "frag"
}
```
```
$ echo "http://john:pass@example.com" | jc -p --url
{ "scheme": "http", "host": "example.com", "user": "john", "password": "pass" }
```
```
$ echo "http://example.com:8080" | jc -p --url
{ "scheme": "http", "host": "example.com", "port": 8080 }
```

**Comment by kellyjonbrazil** [comment]
I like that idea! I’ll look into it. Thanks for the suggestion!

**Comment by kellyjonbrazil** [comment]
I have a version to test ready in the Dev branch. Could you copy [this file](https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/url.py) to your `jc` [plugins folder](https://github.com/kellyjonbrazil/jc#custom-parsers) and let me know how it works for you?

Thanks!

**Comment by kellyjonbrazil** [comment]
This makes me think it might be useful to have `jc` parsers for other common string-types that can be a pain to parse in scripts. Maybe email addresses? I'll see if I can think of any others.

**Comment by archiif** [comment]
Tried it out and it seems to work out pretty nicely!
For the path, do you think it's better for the tool to split it automatically, or just provide it as it is?
Maybe we can also add an extra key called `parts` like in the `yarl` library.
Also, I'm not sure if it's currently just a placeholder, but it's probably better to just use `urlparse.parse_qs` to parse the query string instead of parsing it manually.

**Comment by kellyjonbrazil** [comment]
Thanks for the tips! Playing around with this some more - not sure if I'm going too crazy with this, but I think there could be some neat convenience fields and multiple ways to present the data depending on the use case.

Here is what I'm kicking around (note, it now works with wrapped or unwrapped URLs):
```
$ echo '<URL:http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar#frag>' | jc --url -p
{
  "quoted": "http://user:pass@example.com:8000/a%20space/b/c/d?q1%3Dfoo+with+space%26q2%3Dbar#frag",
  "unquoted": "http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar#frag",
  "scheme": "http",
  "netloc": "user:pass@example.com:8000",
  "path": "/a space/b/c/d",
  "path_list": [
    "a space",
    "b",
    "c",
    "d"
  ],
  "query": {
    "q1": [
      "foo with space"
    ],
    "q2": [
      "bar"
    ]
  },
  "query_list": [
    [
      "q1",
      "foo with space"
    ],
    [
      "q2",
      "bar"
    ]
  ],
  "fragment": "frag",
  "username": "user",
  "password": "pass",
  "hostname": "example.com",
  "port": 8000
}

$ echo "http://user:pass@example.com:8000/a%20space/b/c/d?q1%3Dfoo+with+space%26q2%3Dbar#frag" | jc --url -p
{
  "quoted": "http://user:pass@example.com:8000/a%2520space/b/c/d?q1%253Dfoo%2Bwith%2Bspace%2526q2%253Dbar#frag",
  "unquoted": "http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar#frag",
  "scheme": "http",
  "netloc": "user:pass@example.com:8000",
  "path": "/a%20space/b/c/d",
  "path_list": [
    "a%20space",
    "b",
    "c",
    "d"
  ],
  "query": null,
  "query_list": null,
  "fragment": "frag",
  "username": "user",
  "password": "pass",
  "hostname": "example.com",
  "port": 8000
}
```
I thought it might make sense to have two versions of the query depending on what works best for the end-user. Also, thought it would be cool to include normalized quoted and unquoted URLs in case the user just wants to encode/decode.

What do you think?

**Comment by kellyjonbrazil** [comment]
> Another thing that might be useful is to also provide encoded/decoded versions of each component

Good idea - yeah I can do that. I'll take a look at the `yarl` API as well.

> Also, can you explain what `query_list` is for?

It's just another style of representing the query. The only advantage it has is in case the order of query parameters makes any difference. I'm not sure if that ever matters, though. I may just take it out and have it in my back pocket as an additional representation if someone has an issue with the regular `query` object.

**Comment by kellyjonbrazil** [comment]
Getting closer?
```
% echo '<URL:http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar#frag>' | jc --url -p
{
  "encoded_url": "http://user:pass@example.com:8000/a%20space/b/c/d?q1%3Dfoo+with+space%26q2%3Dbar#frag",
  "unencoded_url": "http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar#frag",
  "scheme": "http",
  "netloc": "user:pass@example.com:8000",
  "path": "/a space/b/c/d",
  "path_encoded": "/a%20space/b/c/d",
  "path_list": [
    "a space",
    "b",
    "c",
    "d"
  ],
  "query": "q1=foo with space&q2=bar",
  "query_encoded": "q1%3Dfoo+with+space%26q2%3Dbar",
  "query_obj": {
    "q1": [
      "foo with space"
    ],
    "q2": [
      "bar"
    ]
  },
  "fragment": "frag",
  "username": "user",
  "password": "pass",
  "hostname": "example.com",
  "port": 8000
}
```

**Comment by archiif** [comment]
I noticed that the query is not validly quoted (the `=` and `&` are quoted)
I guess you can do something like `urllib.parse.quote(string, safe='&=')` instead.
Also, I noticed that in `yarl`, it uses its own [custom quoting function](https://github.com/aio-libs/yarl/blob/master/yarl/_quoting_py.py) for each [compenents](https://github.com/aio-libs/yarl/blob/6b59251623d53252cecb7b521160e77b7bb8d371/yarl/_url.py#L127) of the URL that is quotable. Not sure what special thing it does, but it probably handles a bunch of miscellaneous edge cases.
Not sure how you feel about adding a dependency, but what do you think about just doing an `import yarl`?

**Comment by kellyjonbrazil** [comment]
Yeah, I don't want to add any dependencies, though if it warrants it I would consider vendoring the library. I'll take a look at the Yarl rules and see if I can get close using the stdlibrary quote function. Thanks for the heads up!

**Comment by kellyjonbrazil** [comment]
After taking a closer look at the RFC and Yarl rules, I think I have come up with some sane encoding rules per URL part.
```
# Best-effort to find safe characters in each URL part. Python
    # urllib.parse.quote will always treat the following as safe: `_.-~`
    # https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote
    #
    # Below are additional safe chars per URL part:

    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.1 (scheme)
    SCHEME_SAFE = '+'

    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.2 (netloc)
    NETLOC_SAFE = '+@:[]'

    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.3 (path)
    PATH_SAFE = '+/@:;,='

    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.4 (query)
    QUERY_SAFE = '+/@:;,=&?'

    # https://datatracker.ietf.org/doc/html/rfc3986#section-3.5 (fragment)
    FRAGMENT_SAFE = '+/@:;,=&?'
```
The rules (for the most part) become more relaxed as you get further down the URL, encompassing more and more (but not all) RFC 3986 `reserved` characters.

I think this should work for now, and if there any issues/exceptions found I can add them to a future release. Let me know if you see anything that doesn't look right. Appreciate your feedback and help on this!

**Comment by kellyjonbrazil** [comment]
Here's how the latest version looks:
```
$ echo '<URL:http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar&q2=baz#frag-frag>' | jc --url -p
{
  "url": "http://user:pass@example.com:8000/a space/b/c/d?q1=foo with space&q2=bar&q2=baz#frag-frag",
  "url_encoded": "http://user:pass@example.com:8000/a%20space/b/c/d?q1=foo+with+space&q2=bar&q2=baz#frag-frag",
  "scheme": "http",
  "scheme_encoded": "http",
  "netloc": "user:pass@example.com:8000",
  "netloc_encoded": "user:pass@example.com:8000",
  "path": "/a space/b/c/d",
  "path_encoded": "/a%20space/b/c/d",
  "path_list": [
    "a space",
    "b",
    "c",
    "d"
  ],
  "query": "q1=foo with space&q2=bar&q2=baz",
  "query_encoded": "q1=foo+with+space&q2=bar&q2=baz",
  "query_obj": {
    "q1": [
      "foo with space"
    ],
    "q2": [
      "bar",
      "baz"
    ]
  },
  "fragment": "frag-frag",
  "fragment_encoded": "frag-frag",
  "username": "user",
  "username_encoded": "user",
  "password": "pass",
  "password_encoded": "pass",
  "hostname": "example.com",
  "hostname_encoded": "example.com",
  "port": 8000,
  "port_encoded": "8000"
}
```

**Comment by kellyjonbrazil** [comment]
Working on this a bit more. I decided to update the output and schema so the parser can cover more use cases. Now the parser will output Normalized, Encoded, and Unencoded versions of the URL and fields. The Encoded and Unencoded fields will be under their own keys to clean it up a bit. Seems to be working nicely.
```
Schema:

    {
      "url":                       string,
      "scheme":                    string,
      "netloc":                    string,
      "path":                      string or null,
      "path_list": [               array or null
                                   string
      ],
      "query":                     string or null,
      "query_obj": {               object or null
        <query-key>: [             array or null
          <query-value>            string             # [0]
        ]
      },
      "fragment":                  string or null,
      "username":                  string or null,
      "password":                  string or null,
      "hostname":                  string or null,
      "port":                      integer or null,  # [1]
      "encoded": {
        "url":                     string,
        "scheme":                  string,
        "netloc":                  string,
        "path":                    string or null,
        "path_list": [             array or null
                                   string
        ],
        "query":                   string or null,
        "fragment":                string or null,
        "username":                string or null,
        "password":                string or null,
        "hostname":                string or null,
        "port":                    integer or null,  # [1]
      },
      "decoded": {
        "url":                     string,
        "scheme":                  string,
        "netloc":                  string,
        "path":                    string or null,
        "path_list": [             array or null
                                   string
        ],
        "query":                   string or null,
        "fragment":                string or null,
        "username":                string or null,
        "password":                string or null,
        "hostname":                string or null,
        "port":                    integer or null,  # [1]
      }
    }

    [0] Duplicate query-keys will have their values consolidated into the
        array of query-values

    [1] Invalid port values will be converted to null/None and a warning
        message will be printed to `STDERR` if quiet=False
```

**Comment by kellyjonbrazil** [comment]
This parser is now released in version 1.20.4.

BTW, this parser inspired me to also create:
- ISO 8601 Datetime string parser
- UNIX Epoch Timestamp string parser
- JWT string parser
- Email Address string parser
